### PR TITLE
[#10969] Search form for audit logs page

### PR DESCRIPTION
**Pull request by t-cheepeng**
Part of #10969

**Outline of solution**
- Added a search form to audit logs page as per wireframe in #10969

**Current Page**
image

### CODE REVIEW

File: src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.html [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        <div class="col-12">
+          <label for="student-name-input">Student Name (Optional)</label>
+        </div>
+        <div class="col-12">
+          <input id="student-name-input" class="form-control col-6" type="text" [(ngModel)]="formModel.studentName" placeholder="Student Name"/>
+        </div>
+      </div>
+      <div class="col-1">
+        <button class="btn btn-primary mt-4" (click)="search()">Search</button>
+      </div>
```
Comment (halfwhole) [inline src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.html:64]: Is there a way to align these with the row above? Currently the 5/6/1 column split in the bottom row looks a bit off compared to the 6/6 in the top row

**Comment by t-cheepeng** [comment]
> 
> 
> Do we need to also add a Feedback Session Name dropdown that is generated after the Course ID is chosen?

I think the plan is to fetch the logs for the whole of the courseid. possibly include the advanced filtering in a future PR

**Review by AdithyaNarayan — APPROVED** [review]
LGTM

File: src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.html [hunk 2]
```diff
@@ -0,0 +1,73 @@
+<h1>Student Activity Logs</h1>
+<p>This page allows you to find when your students have accessed or submitted a particular feedback session.</p>
```
Comment (madanalogy) [inline src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.html:2]: Are there plans to include filtering based on feedback session after selecting / filling in the course ID? If yes, then you might want to include a skeleton template for that. If no, then maybe reword this prompt to include "for a given course"

File: src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+  ngOnInit(): void {
+    this.loadCourses();
+  }
+
+  /**
+   * Search for logs of student activity
+   */
+  search(): void {
+    // TODO: Call endpoint to retrieve logs
```
Comment (madanalogy) [inline src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.ts:54]: Might be more worthwhile to have this PR on hold until the API endpoint and data formats are finalised

File: src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.html [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      </div>
+    </div>
+    <div class="row form-group">
+      <div class="col-6">
+        <div class="col-12">
+          <label for="course-id-dropdown">Course ID</label>
+        </div>
+        <div class="col-12">
+          <select id="course-id-dropdown" class="form-control col-6" [(ngModel)]="formModel.courseId">
+            <option *ngFor="let course of courses">
```
Comment (madanalogy) [inline src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.html:48]: Not sure how this would work out for instructors with lots of courses. Any possibility of having the courses sorted by date or allowing filtering based on text input? Or is that already a built in function of the `select` field?

**Review by madanalogy — COMMENTED** [review]
UI looks good, just some comments below

File: src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.ts (hunk 3 above)
Comment (t-cheepeng) [reply src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.ts:54]: The endpoint return type is specified in #10978 but should not really matter for this PR since the PR only deals with the search form for now

File: src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.html (hunk 2 above)
Comment (t-cheepeng) [reply src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.html:2]: Will reword it

File: src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.html (hunk 4 above)
Comment (t-cheepeng) [reply src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.html:48]: How about sort by alphabetical order? I would think that most instructors don't really have that many courses that the dropdown will impede the UX?

Comment (madanalogy) [reply src/web/app/pages-instructor/instructor-audit-logs-page/instructor-audit-logs-page.component.html:48]: > I would think that most instructors don't really have that many courses that the dropdown will impede the UX?

Best to check UI/UX matters with @damithc

**Comment by damithc** [comment]
Regrading the UI:
Better term for `Logs from`? e.g., `Search period: from, until`
The drop-down for the time is too wide? 
I'm interested to see how the search results will be displayed.

On a separate note, is the backend capable of searching the full period within the 1-minute window given to each request?

**Comment by t-cheepeng** [comment]
@damithc 
Have shortened the time picker and bolded/changed the form names. Is the search form okay? 
image

>I'm interested to see how the search results will be displayed.

That should be coming in a separate PR from this, but shouldn't look too dissimilar from the wireframe 

>On a separate note, is the backend capable of searching the full period within the 1-minute window given to each request?

From some preliminary testing on the staging server, retrieving 2000 log entries without filtering (about 2 days worth) from request logs took about 1 minute. With more conditions on the retrieval, the period of time searched for can be longer. But it is unlikely that we are able to stay below 1 minute beyond 5 days worth of logs (from a rough estimate). Will need more testing on this

**Comment by damithc** [comment]
> Have shortened the time picker and bolded/changed the form names. Is the search form okay?

It's better now. Note that layout (i.e., alignment, spacing) can be used to convey additional information about the meaning of the fields. 
image
The current layout suggests two main groups of inputs but that grouping is not aligned with the actual meaning of the inputs.
Also, the submit button is a bit out of place.



> From some preliminary testing on the staging server, retrieving 2000 log entries without filtering (about 2 days worth) from request logs took about 1 minute. With more conditions on the retrieval, the period of time searched for can be longer. But it is unlikely that we are able to stay below 1 minute beyond 5 days worth of logs (from a rough estimate). Will need more testing on this

Got it. That means the output could be _up to_ a certain end time which is earlier than the actual end time specified by the user. The UI for the results should indicate that clearly.

**Comment by t-cheepeng** [comment]
@damithc 
Two different groupings shown below. Which one would be clearer? 

Grouping 1:
image

Grouping 2:
Screenshot 2021-03-15 111246

**Comment by damithc** [comment]
> Two different groupings shown below. Which one would be clearer?

Of these two, I prefer the latter. You can also get some ideas from the form used to create a session.

* If you think it is more natural for the user to choose the course first, that should be on the left half.
* Is the user expected to enter the student name exactly as it is in the database? that's not very realistic. Perhaps it should be a dropdown instead?

**Comment by t-cheepeng** [comment]
@damithc Swapped the course id and student to the left side. definitely makes more sense there. Also, the student is a dropdown now
image

**Comment by damithc** [comment]
> @damithc Swapped the course id and student to the left side. definitely makes more sense there. Also, the student is a dropdown now
> image

Looks good enough to me. 👍

**Review by rrtheonlyone — APPROVED** [review]
LGTM 👍
